File: backend/app/services/storage_service.py
from __future__ import annotations

import re
import uuid
from pathlib import Path

from fastapi import HTTPException, UploadFile, status

from app.core.config import settings


class StorageService:
    def __init__(self) -> None:
        self.root = Path(settings.media_root)

    async def save_upload(
        self,
        file: UploadFile,
        category: str,
        *,
        allowed_content_types: set[str],
        max_size_bytes: int,
    ) -> dict[str, str | int]:
        content_type = file.content_type or "application/octet-stream"
        if content_type not in allowed_content_types:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Unsupported file type: {content_type}",
            )

        original_name = file.filename or f"{category}-upload"
        safe_name = self._sanitize_filename(original_name)
        target_dir = self.root / category
        target_dir.mkdir(parents=True, exist_ok=True)
        target_path = target_dir / f"{uuid.uuid4()}-{safe_name}"

        size = 0
        with target_path.open("wb") as output:
            while True:
                chunk = await file.read(1024 * 1024)
                if not chunk:
                    break
                size += len(chunk)
                if size > max_size_bytes:
                    target_path.unlink(missing_ok=True)
                    raise HTTPException(
                        status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                        detail="Uploaded file exceeds the size limit.",
                    )
                output.write(chunk)

        await file.seek(0)

        return {
            "name": original_name,
            "file_path": str(target_path),
            "content_type": content_type,
            "size_bytes": size,
        }

    def delete_file(self, file_path: str | None) -> None:
        if not file_path:
            return
        path = Path(file_path)
        try:
            path.unlink(missing_ok=True)
        except OSError:
            return

    def _sanitize_filename(self, name: str) -> str:
        sanitized = re.sub(r"[^A-Za-z0-9._-]+", "-", name).strip("-")
        return sanitized or "upload"


storage_service = StorageService()

File: core/storage/storageService.ts
import { createWriteStream } from "node:fs";
import { mkdir, unlink } from "node:fs/promises";
import { join, resolve } from "node:path";
import { randomUUID } from "node:crypto";
import { Readable } from "node:stream";
import { finished } from "node:stream/promises";

import { getConfig } from "../config/index.js";
import { badRequest, payloadTooLarge } from "../errors.js";

export interface SavedUpload {
  name: string;
  filePath: string;
  contentType: string;
  sizeBytes: number;
}

export interface SaveUploadOptions {
  allowedContentTypes: Set<string>;
  maxSizeBytes: number;
}

/** Keep uploaded filenames filesystem-safe while preserving a readable suffix. */
function sanitizeFilename(name: string): string {
  const sanitized = name.replace(/[^A-Za-z0-9._-]+/g, "-").replace(/^-+|-+$/g, "");
  return sanitized || "upload";
}

export class StorageService {
  private readonly root: string;

  constructor() {
    this.root = resolve(getConfig().mediaRoot);
  }

  async saveUpload(file: File, category: string, options: SaveUploadOptions): Promise<SavedUpload> {
    const contentType = file.type || "application/octet-stream";
    if (!options.allowedContentTypes.has(contentType)) {
      throw badRequest(`Unsupported file type: ${contentType}`);
    }

    const originalName = file.name || `${category}-upload`;
    const safeName = sanitizeFilename(originalName);
    const targetDir = join(this.root, category);
    await mkdir(targetDir, { recursive: true });
    const targetPath = join(targetDir, `${randomUUID()}-${safeName}`);

    let size = 0;
    const writeStream = createWriteStream(targetPath);
    const nodeStream = Readable.fromWeb(file.stream() as never);

    try {
      for await (const chunk of nodeStream) {
        size += chunk.length;
        if (size > options.maxSizeBytes) {
          writeStream.destroy();
          await unlink(targetPath).catch(() => undefined);
          const limitMb = Math.floor(options.maxSizeBytes / (1024 * 1024));
          throw payloadTooLarge(`Uploaded file exceeds the configured size limit of ${limitMb} MB.`);
        }
        if (!writeStream.write(chunk)) {
          await new Promise<void>((resolveDrain) => writeStream.once("drain", () => resolveDrain()));
        }
      }
      writeStream.end();
      await finished(writeStream);
    } catch (error) {
      writeStream.destroy();
      await unlink(targetPath).catch(() => undefined);
      throw error;
    }

    return { name: originalName, filePath: targetPath, contentType, sizeBytes: size };
  }

  async deleteFile(filePath: string | null | undefined): Promise<void> {
    if (!filePath) return;
    await unlink(filePath).catch(() => undefined);
  }
}

export const storageService = new StorageService();
